design/1.0.md

```diff
@@ -0,0 +1,84 @@
+
+# 1.0 Targets
+
+## Basic taint propagation detection
+
+Given a spec defining sources and sinks (below), diagnostics are reported for
+any direct call to a sink that includes a source.  Additionally, we detect any
+taint propagation from a source to a sink within the scope of a single function.
+
+## Analysis Configuration
+
+Configuration will consist of the following key components:
+
+* Identification of sources, sinks, and sanitizers
+* Identification of analysis scope, e.g. to skip third-party or testing packages.
+* Specify diagnostics to be ignored, e.g false-positives
+
+### Identification of sources, sinks, and sanitizers
+
+Much of our work is directly with `ssa.Value`s and `ssa.Instruction`s.
+Whether those values or instructions represent one element or another,
+ the configuration to identify these members should be as uniform as possible.
+In that light, specification should follow:
+
+#### ValueSpecifier
+A ValueSpecifier marks an `ssa.Value`, e.g., for identification as a source or as the safe result of a sanitizer.
+Values should be specifiable via any or all of:
+
+* Type / Field: Package path and type name, optional field name
+* Field tags
+* Scope: local variable, free variable, or global
+* Context: within a specific package or function call.  See CallSpecifier below.
+* Is a reference
+* Is "reference-like," e.g., slices, maps, chans.
+
+#### CallSpecifier
+A CallSpecifier marks an `*ssa.Call`, e.g., as a sink or sanitizer.
+Calls should be specifiable via any or all of:
+
+* Symbol: package path, function name, optional receiver name.
+* Context: A CallSpecifier indicating  
+* Based on arguments (ValueSpecifier)
+* Based on return values (ValueSpecifier)
+
+## Inference
+
+Inference represents the scope of what a user does *not* need to configure.
+For instance, we can currently detect getter methods and mark as propagating
+a marked field without requiring any explicit configuration.
+
+If a source type is aliased, the alias is properly identified as equivalent as the aliased source type.
+
+# Beyond a 1.0
+
+The following seem like significant improvements but seem larger in scope than what constitutes a stable 1.0.
+These items are open to interpretation and warrant discussion.
+
+## Improved Inference
+
+* Detect sanitization via zero-value.
+    * Explicit sanitization of `type Source struct{ data string }` is easily done via `s.data = ""`.
+    * Alternatively, a user may specify redaction placeholders, e.g. `"REDACTED"` rather than require a zero value.
+* A type defined from a source type (e.g., `type Foo core.Source`) should also be considered a source type.
+    * Note: as tags take part in type identify of the underlying struct literal, field tags are "inherited" in defined types based on existing types.
+      This may result in conflicting current behavior between those sources identified by tag and those identified by path/package/name.
```

Comment:
I think I understand what you mean, but could you give an example here?

Reply:
I'd need to test it, but I believe that...

```golang
type SourceA struct {
  Data string    // specified in config by package, name, and field
}

type SourceB struct {
  Data string  // `levee:"source"`
}

type NamedA SourceA  // Not identified as a source

type NamedB SourceB  // Is identified by source
```

Since the underlying struct for both `SourceB` and `NamedB` is the struct literal
```golang
struct {
  Data string // `levee:"source"`
}```

then both will satisfy the identification by field tags.  But since the other is identified by name-matching regexp, the change of name will lost track of it.

Both of these should be handled uniformly by the `sourceType` analyzer (which also need to actually be consumed by the `source` analyzer, and either be unified with or consume the output of the `fieldtags` analyzer).

Reply:
Good explanation, thanks! This will be something to keep in mind then.